speedup light effects by using shared material for drones

vasarhelyi opened the pull request:
This PR changes a fundamental concept in how colors are stored for Skybrush drone shows: we do _not_ have a detached material for each drone any more, but have a common template material only, and use a different shader node tree that inputs object.color into the node tree. This should speed up color and light effects handling and export substantially! The PR is not ready, TODOs are summarized below:

- [x] users should be notified to change Solid Viewport Shading mode's "Object Color" to "Object" by default, to see the colors properly. In script: `bpy.context.space_data.shading.color_type = 'OBJECT'`
- [x] a migration operator needs to be created to convert animation of old shows to new workflow and users should be asked automatically whether they wish to convert a show when they open an old show
- [x] now I create an action for each drone for its color animation, but I have not checked whether it gets into conflict with other actions we might have created previously for other reasons. @ntamas can you comment on this?
- [x] once all features above are ready, old code that is not used any more related to materials can be removed
- [x] documentation needs to be updated
- [x] general code review and testing

ntamas commented on src/modules/sbstudio/plugin/operators/migrations/use_common_material_for_all_drones.py:
```diff
@@ -0,0 +1,149 @@
+def set_all_shading_color_types_to_object() -> None:
+    """Sets the object color source of the solid and wireframe
+    shading types of the 3D Viewport to use object color."""
+    shading = getattr(bpy.context.space_data, "shading", None)
```
`bpy.context.space_data` might be a completely different object depending on how the workspace of the user is set up. We need to find the 3D viewport with `find_one_3d_view()` or `find_all_3d_views()` and then work on those.

vasarhelyi replied:
done

vasarhelyi commented on src/modules/sbstudio/plugin/tasks/initialization.py:
```diff
@@ -71,6 +71,21 @@ def update_pyro_particles_of_drones(*args):
         update_pyro_particles_of_object(drone)
 
 
+def config_logging(*args):
```
done, thx

vasarhelyi commented:
Removed dead code and added documentation, I think I am done with this PR. 

@ntamas note that this PR brings major change, so version 4.0 (and internal version=2) is hardcoded in it at many places, so if we merge to main we need to release 4.0 soon as the next version. There is only one more pending PR there that could fit into version 4.0, this: https://github.com/skybrush-io/studio-blender/pull/40, we can wait with the merge until that is merged, or any order that is convenient... And perhaps there is the new licensing part that is still under development, we could wait for that as well with version 4.0!

ntamas commented on doc/modules/ROOT/pages/concepts.adoc:
```diff
@@ -101,13 +101,28 @@ NOTE: While using the regular return to home, it is not guaranteed that each dro
 
+=== Storage and visualization of color data
+
+Baked colors of a given drone object is stored in the  animation data of the object's own `color` property, while parametric light effect colors are dynamically overlaid onto this property for the current frame in real-time.
```
```suggestion
Baked colors of a given drone object are stored in the animation data of the `color` property of the object (shown as "Color" in the "Viewport Display" section in the Properties editor), while parametric light effect colors are dynamically overlaid onto this property for the current frame in real-time.
```

ntamas commented on doc/modules/ROOT/pages/concepts.adoc:
```diff
@@ -101,13 +101,28 @@ NOTE: While using the regular return to home, it is not guaranteed that each dro
+
+Baked colors of a given drone object is stored in the  animation data of the object's own `color` property, while parametric light effect colors are dynamically overlaid onto this property for the current frame in real-time.
+
+For properly visualizing colors, internally we use a common drone template material for all drones (which should always be the first material of each drone) with a shader node tree that inputs object color as the color source for the material. Thus for properly visualizing drone colors, you should set the "Object Color" and/or "Wireframe Color" properties of the Viewport Shading panel in your 3D  ...
```
```suggestion
For properly visualizing colors, internally we use a common drone template material for all drones (which should always be the first material of each drone) with a shader node tree that injects the viewport color of the object (i.e. the `color` property) as the color source for the material. Therefore, for properly visualizing drone colors, you should set the "Object Color" and/or "Wireframe Color" properties of the Viewport Shading panel in your 3D viewports to `Object`.
```

ntamas commented on doc/modules/ROOT/pages/concepts.adoc:
```diff
@@ -101,13 +101,28 @@ NOTE: While using the regular return to home, it is not guaranteed that each dro
+
+For properly visualizing colors, internally we use a common drone template material for all drones (which should always be the first material of each drone) with a shader node tree that inputs object color as the color source for the material. Thus for properly visualizing drone colors, you should set the "Object Color" and/or "Wireframe Color" properties of the Viewport Shading panel in your 3D  ...
+
+Note that earlier (pre 4.0) versions of the *Skybrush Studio for Blender* plugin used a different shader node tree with one unique material for each drone and storing color animation in these individual materials (we dropped this solution due to efficiency constraints). So when you open an old Blender file, a migration window should pop up warning you about the deprecated structure of your old fi ...
```
```suggestion
Note that earlier (pre 4.0) versions of the *Skybrush Studio for Blender* plugin used a different shader node tree with one unique material for each drone. This solution was changed in version 4.0 due to efficiency reasons. When you open a file that was created with an earlier version of the plugin, a migration window should pop up warning you about the deprecated structure of your old file and providing means for upgrading your file to the new structure automatically with a single click.
```

ntamas reviewed (changes requested):
Some doc changes as suggested above; feel free to accept or ignore. I'll merge this when the time comes for 4.0.

vasarhelyi commented:
docs suggestions all accepted, thx!
